core/src/main/java/org/infinispan/container/offheap/OffHeapDataContainer.java

```diff
@@ -345,8 +345,8 @@ protected void entryRetrieved(long entryAddress) {
             size.decrementAndGet();
             return ice;
          }
-         prevAddress = address;
-         address = nextAddress;
+         prevAddress = bucketHeadAddress;
+         bucketHeadAddress = nextAddress;
```

Comment:
One more small complaint: once you change `bucketHeadAddress` it's no longer a the head of the list, so you should use a local variable for iteration.

Reply:
Sure I can just assign it to address on the first line instead.